### PR TITLE
refactor(error): remove unnecessary error return

### PR DESCRIPTION
**Pull request by BlackHole1**
If a function does not produce any errors, it should not return an error (except for functions defined by an interface). This will reduce the mental burden on users.

The AddDevices method was also added, on the side.

### CODE REVIEW

**Comment by gbraad** [comment]
Thank you. I'll let @cfergeau decide as there might have been reasons, though he is currently on leave.

**Comment by gbraad** [comment]
> The AddDevices method was also added, on the side. (If you are concerned about this, I will move it to another PR.)

Knowing Christophe, this should have been at least a separate commit, but best to make a PR specific to this with clear description.

**Comment by gbraad** [comment]
Waiting for repush to ok-for-test

**Comment by BlackHole1** [comment]
> this should have been at least a separate commit

Done

File: pkg/config/json_test.go [hunk 1]
```diff
@@ -62,55 +55,38 @@ var jsonTests = map[string]jsonTest{
...
+			vm.AddDevice(dev)
 			// virtio-gpu
-			dev, err = VirtioGPUNew()
-			require.NoError(t, err)
-			err = vm.AddDevice(dev)
+			dev = VirtioGPUNew()
 			require.NoError(t, err)
+			vm.AddDevice(dev)
 			// virtio-net
 			dev, err = VirtioNetNew("00:11:22:33:44:55")
```
Comment (cfergeau) [inline pkg/config/json_test.go:69]: This is the main reason why most `*New()` methods return an error, some of them require an error return, so for me it's better if they are consistent and all return an error, instead of having to remember it. This also makes the API future-proof if one day we need to return an error.

File: pkg/config/config.go [hunk 2]
```diff
@@ -142,10 +142,8 @@ func (vm *VirtualMachine) VirtioVsockDevices() []*VirtioVsock {
 
 // AddDevice adds a dev to vm. This device can be created with one of the
 // VirtioXXXNew methods.
-func (vm *VirtualMachine) AddDevice(dev VirtioDevice) error {
+func (vm *VirtualMachine) AddDevice(dev VirtioDevice) {
 	vm.Devices = append(vm.Devices, dev)
```
Comment (cfergeau) [inline pkg/config/config.go:146]: Yep, this one is simpler without the error.

**Review by cfergeau — COMMENTED** [review]
Sorry for the late reviews, you filed these PRs just before I went away for a while :)
I'll need to think some more about the error removal from the `*New` changes, there are some plus and minuses to having it/not having it :)

File: pkg/config/config.go [hunk 3]
```diff
@@ -146,6 +146,11 @@ func (vm *VirtualMachine) AddDevice(dev VirtioDevice) {
...
 }
 
+// AddDevices adds a list of devices to vm.
```
Comment (cfergeau) [inline pkg/config/config.go:149]: This is what `AddDevice` should have been ;) maybe change it to `return AddDevices(dev)`

Comment (BlackHole1) [reply pkg/config/config.go:149]: Good idea!

**Comment by BlackHole1** [comment]
Hey @cfergeau. I agree with what you said: maintaining a consistent style is also a way to reduce cognitive load.

In this current PR, I noticed that there are only two `FooNew()` methods that return an error, and I’m wondering if we could also make them not return an error to maintain a consistent style.

1. `VirtioInputNew`: This method calls `dev.validate()`, but it actually doesn’t need to because other `FooNew()` methods will call `dev.validate()` during the `ToCmdLine` and `FromOptions` stages. So it’s pretty simple to make this method *no-err*.
2. `VirtioNetNew`: This method needs to call `net.ParseMAC` to determine whether the input mac address is correct. If we change the function signature to: `VirtioNetNew(macAddress net.HardwareAddr)`, could we avoid having this function return an error?

-----

as you said: "This also makes the API future-proof if one day we need to return an error". is also a reason. But for me: The optimal design for now will surely not be optimal in the future, avoid over-design. When the future returns an error, we can consider ways to improve, such as changing the function name to: `FooSetup`. At this stage, `FooNew` only does one thing, which is to create a struct instance.

File: pkg/rest/vf/vm_config.go [hunk 4]
```diff
@@ -18,7 +18,7 @@ func NewVzVirtualMachine(vm *vz.VirtualMachine, config *vz.VirtualMachineConfigu
 	return &VzVirtualMachine{config: config, VzVM: vm}
 }
 
-// inspect returns information about the virtual machine like hw resources
+// Inspect returns information about the virtual machine like hw resources
```
Comment (cfergeau) [inline pkg/rest/vf/vm_config.go:21]: Can you split these comment changes in a separate commit?

**Comment by cfergeau** [comment]
> as you said: "This also makes the API future-proof if one day we need to return an error". is also a reason. But for me: The optimal design for now will surely not be optimal in the future, avoid over-design. When the future returns an error, we can consider ways to improve, such as changing the function name to: `FooSetup`. At this stage, `FooNew` only does one thing, which is to create a struct instance.

For internal/private API, I agree. For external API, changing the API in incompatible ways has to be balanced with the annoyances this will cause to users.
Removing the errors mean changing https://github.com/crc-org/crc/blob/2b661eb9eb2c57676e1f18b728e8aa5c025ba360/pkg/drivers/vfkit/driver_darwin.go#L187-L196 , changing https://github.com/containers/podman/blob/01b2243e73076a0e71a667153af1b98ecbf3b2f5/pkg/machine/applehv/vfkit.go#L10-L33

And I don't know if there are other users of vfkit's config code.
If we make the change now, and have to change it back in the future to add the error, then we'll have broken our API twice, and annoyed twice vfkit go users. This is why I'm very conservative with these changes. If we don't have a choice, I'll consider it, otherwise, I'll think hard about it before making the change.

(and API changes get even harder after the module has had a v1 release, as breaking API means we have to change the module path, which causes even more churn to go programs using vfkit)

**Comment by BlackHole1** [comment]
@cfergeau Okay, you convinced me. I will close this PR and create two new ones to address the following issues:

1. Add the `AddDevices` method
2. Improve comments

**Comment by cfergeau** [comment]
> @cfergeau Okay, you convinced me. I will close this PR and create two new ones to address the following issues:
> 
>     1. Add the `AddDevices` method
> 
>     2. Improve comments

Sharing our different point of views made for an interesting discussion! Thanks for the 2 additional PRs :)
